File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/scenario/given/steps/BaseGivenStep.java
package com.mycompany.myshop.testkit.dsl.core.scenario.given.steps;

import com.mycompany.myshop.testkit.dsl.core.usecase.UseCaseDsl;
import com.mycompany.myshop.testkit.dsl.core.scenario.given.GivenImpl;
import com.mycompany.myshop.testkit.dsl.port.then.ThenStage;
import com.mycompany.myshop.testkit.dsl.port.given.steps.base.GivenStep;
import com.mycompany.myshop.testkit.dsl.core.scenario.when.WhenImpl;

public abstract class BaseGivenStep implements GivenStep {
    private final GivenImpl given;

    protected BaseGivenStep(GivenImpl given) {
        this.given = given;
    }

    public GivenImpl and() {
        return given;
    }

    public WhenImpl when() {
        return given.when();
    }

    public ThenStage then() {
        return given.then();
    }

    public abstract void execute(UseCaseDsl app);
}

File: system-test/dotnet/Dsl.Core/Scenario/Given/Steps/Base/BaseGivenStep.cs
using Dsl.Core.Scenario.Then;
using Dsl.Core.Scenario.When;
using Dsl.Port.Given;
using Dsl.Port.Given.Steps.Base;
using Dsl.Port.Then;
using Dsl.Port.When;
using Driver.Adapter;
using Optivem.Testing;

namespace Dsl.Core.Scenario.Given;

public abstract class BaseGiven : IGivenStep
{
    private readonly GivenStage _givenClause;

    protected BaseGiven(GivenStage givenClause)
    {
        _givenClause = givenClause;
    }

    public GivenStage And()
    {
        return _givenClause;
    }

    IGivenStage IGivenStep.And() => And();

    public WhenStage When()
    {
        return _givenClause.When();
    }

    IWhenStage IGivenStep.When() => When();

    public ThenStageBase Then()
    {
        return _givenClause.Then();
    }

    IThenStage IGivenStep.Then() => Then();

    internal abstract Task Execute(UseCaseDsl app);

    protected Channel? Channel => _givenClause.Channel;
}


